### PR TITLE
[dicp]Cjt/support resnet50 train

### PR DESCRIPTION
**Pull request by KevinfromTJ**
support training resnet50 in ascend

### CODE REVIEW

**Comment by KevinfromTJ** [comment]
TODO：需要往华为ci里加resnet50环境

**Review by zhaochaoxing — APPROVED** [review]

File: dicp/test/model/test_resnet50.py [hunk 1]
```diff
@@ -76,8 +77,10 @@ def test_forward_train(self, backend, dynamic, fake_batch_num, batch_size):
...
             torch_dipu.current_stream().synchronize()
             dicp_real_loss = dicp_loss["loss"]
+            print(i,"--DICP: ", dicp_real_loss)
             dicp_optimizer.zero_grad()
             dicp_real_loss.backward()
             dicp_optimizer.step()
 
-            assert torch.allclose(cpu_real_loss.detach(), dicp_real_loss.cpu().detach(), atol=1e-01, equal_nan=True)
+            # assert torch.allclose(cpu_real_loss.detach(), dicp_real_loss.cpu().detach(), atol=1e-01, equal_nan=True)
+            assert torch.allclose(cpu_real_loss.repeat(1).detach(), dicp_real_loss.repeat(1).cpu().detach(), atol=1e-01, equal_nan=True)
```
Comment (jinminxi104) [inline dicp/test/model/test_resnet50.py:86]: repeat(1)为了什么？

File: dicp/dicp/vendor/AscendGraph/conversion.py [hunk 2]
```diff
@@ -794,7 +794,7 @@ def _native_batch_norm_legit_functional(self, x, weight, bias, running_mean, run
                                     bn_reduce, 1, weight, bias,
                                     running_mean, running_var, eps, momentum))
         # 3. tie all results: result, saved_mean, saved_invstd
-        edges = {bn_update.node.name: [
+        edges = {bn_update.node.name + "_edge_name": [
```
Comment (jinminxi104) [inline dicp/dicp/vendor/AscendGraph/conversion.py:797]: 添加“edge_name”的目的是什么？

**Review by jinminxi104 — CHANGES_REQUESTED** [review]

**Comment by KevinfromTJ** [comment]
1.华为这边dicp和ascend输出好像一个标量一个是1维张量，torch allclose报错shape '[1]'和'[]'无法广播？这个当时没找到是不是其他地方的问题，repeat(1)把标量变成1维张量
不知道遂原那边输出的loss是什么样的，有没有这种问题
2.这个得问赵老师，个人理解是batchnorm中有一些参数和要保存的量，像IdentityN这种算子会连到后续五个变量的Identity算子形成一个图或者树的结构，可能是华为接口要求也可能是自己定义的，当时出错的原因就是和codegen的ascend.py里面的函数没统一命名，我只是把它们都改过来
&nbsp;


kevin
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2024年1月23日(星期二) 凌晨0:11
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [DeepLink-org/deeplink.framework] [dicp]Cjt/support resnet50 train (PR #635)





 
@jinminxi104 requested changes on this pull request.
 
 
In dicp/test/model/test_resnet50.py:
 &gt;              dicp_optimizer.zero_grad()              dicp_real_loss.backward()              dicp_optimizer.step()   -            assert torch.allclose(cpu_real_loss.detach(), dicp_real_loss.cpu().detach(), atol=1e-01, equal_nan=True) +            # assert torch.allclose(cpu_real_loss.detach(), dicp_real_loss.cpu().detach(), atol=1e-01, equal_nan=True) +            assert torch.allclose(cpu_real_loss.repeat(1).detach(), dicp_real_loss.repeat(1).cpu().detach(), atol=1e-01, equal_nan=True)  
repeat(1)为了什么？
 
 
In dicp/dicp/vendor/AscendGraph/conversion.py:
 &gt; @@ -794,7 +794,7 @@ def _native_batch_norm_legit_functional(self, x, weight, bias, running_mean, run                                      bn_reduce, 1, weight, bias,                                      running_mean, running_var, eps, momentum))          # 3. tie all results: result, saved_mean, saved_invstd -        edges = {bn_update.node.name: [ +        edges = {bn_update.node.name + "_edge_name": [  
添加“edge_name”的目的是什么？

File: dicp/dicp/vendor/AscendGraph/conversion.py (hunk 2 above)
Comment (jinminxi104) [reply dicp/dicp/vendor/AscendGraph/conversion.py:797]: talked offline

**Review by jinminxi104 — APPROVED** [review]
